mounts.go

```diff
@@ -492,54 +507,82 @@ func (f *FileMount) Mount() error {
 	// Let's copy rather than move, since we want to be totally sure that the
 	// task can't modify the contents, and setting as read-only is not enough -
 	// the user could change the rights and then modify it.
+	task.Infof("[mounts] Copying cached version of %v to %v", fsContent.String(), file)
 	err = copyFileContents(cacheFile, file)
 	if err != nil {
-		panic(fmt.Errorf("Could not copy file %v to %v due to:\n%v", cacheFile, file, err))
+		// this could be a system error, but it can also be that e.g. the task
+		// specified an invalid path, so resolve as malformed payload rather
+		// than panic
+		task.Errorf("[mounts] Not able to mount content from %v at path %v", fsContent.String(), file)
+		task.Infof("%v", err)
+		return err
 	}
 	return nil
 }
 
 // Nothing to do - original archive file was copied, not moved
-func (f *FileMount) Unmount() error {
+func (f *FileMount) Unmount(task *TaskRun) error {
 	return nil
 }
 
 // ensureCached returns a file containing the given content
-func ensureCached(fsContent FSContent) (file string, err error) {
+func ensureCached(fsContent FSContent, task *TaskRun) (file string, err error) {
 	cacheKey := fsContent.UniqueKey()
+	var sha256 string
 	if _, inCache := fileCaches[cacheKey]; !inCache {
-		file, err := fsContent.Download()
+		// fsContent.Download already takes care of ensuring SHA256 is correct,
+		// so don't need to do that here
+		file, sha256, err = fsContent.Download(task)
 		if err != nil {
 			log.Printf("Could not download %v due to %v", fsContent.UniqueKey(), err)
-			return "", err
+			return
 		}
 		fileCaches[cacheKey] = &Cache{
 			Location: file,
 			Hits:     1,
 			Created:  time.Now(),
 			Owner:    fileCaches,
 			Key:      cacheKey,
+			SHA256:   sha256,
 		}
-	} else {
-		// Sanity check - if file is in file map, but not on file system,
-		// something is seriously wrong, so should be a worker exception
-		// (panic), not a task failure
-		_, err := os.Stat(fileCaches[cacheKey].Location)
-		if err != nil {
-			panic(fmt.Errorf("File in cache, but not on filesystem: %v", *fileCaches[cacheKey]))
+		err = validateSHA256(fsContent.String(), sha256, fsContent.RequiredSHA256(), task)
```

Comment:
Is this unhandled `err` result a bug?

Reply:
No, in this case, it is a [named result parameter](https://golang.org/doc/effective_go.html#named-results), which will is handled by calling code.